device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,252 @@
+     *
+     * @param protocol The transport protocol that this client will build the multiplexed connection on. Must be either
+     *                 {@link IotHubClientProtocol#AMQPS} or {@link IotHubClientProtocol#AMQPS_WS}.
+     */
+    public MultiplexingClient(IotHubClientProtocol protocol)
+    {
+        this(protocol, null);
+    }
+
+    /**
+     * Instantiate a new MultiplexingClient that will establish a multiplexed connection through a proxy.
+     *
+     * @param protocol The transport protocol that this client will build the multiplexed connection on. Must be
+     * {@link IotHubClientProtocol#AMQPS_WS} since using {@link IotHubClientProtocol#AMQPS} does not support proxies.
+     * @param proxySettings The proxy settings that this client will use.
+     */
+    public MultiplexingClient(IotHubClientProtocol protocol, ProxySettings proxySettings)
+    {
+        switch (protocol)
+        {
+            case AMQPS:
+            case AMQPS_WS:
+                break;
+            default:
+                throw new IllegalArgumentException("Multiplexing is only supported for AMQPS and AMQPS_WS");
+        }
+
+        this.deviceClientList = new ArrayList<>();
+        this.protocol = protocol;
+        this.proxySettings = proxySettings;
+    }
+
+    /**
+     * Opens this multiplexing client. At least one device client must be registered prior to calling this.
+     * <p>
+     * This call behaves synchronously, so if it returns without throwing, then all registered device clients were
+     * successfully opened.
+     * <p>
+     * If this client is already open, then this method will do nothing.
+     * <p>
+     * @throws IOException If any IO errors occur while opening the multiplexed connection.
+     */
+    public void open() throws IOException
+    {
+        synchronized (this.lock)
+        {
+            if (deviceClientList.size() < 1)
+            {
+                throw new UnsupportedOperationException("Must register at least one device client before opening a multiplexed connection");
+            }
+
+            log.info("Opening multiplexing client");
+            this.deviceIO.open();
+            log.info("Successfully opened multiplexing client");
+        }
+    }
+
+    /**
+     * Close this multiplexing client. This will close all active device sessions as well as the AMQP connection.
+     * <p>
+     * If this client is already closed, then this method will do nothing.
+     * <p>
+     * Once closed, this client can be re-opened. It will preserve the previously registered device clients.
+     * <p>
+     * @throws IOException If any exception occurs while closing the connection.
+     */
+    public void close() throws IOException
+    {
+        synchronized (this.lock)
+        {
+            log.info("Closing multiplexing client");
+            for (DeviceClient deviceClient : this.deviceClientList)
+            {
+                deviceClient.closeFileUpload();
```

Comment:
Seems this might take a long time. No way to parallelize?

Reply:
Not without spawning other threads, which is a bit much since we should aim to support constrained devices

Reply:
I guess we could make it an option (even later), of how many threads to handle parallelization across the clients. Default would be 1, which would maintain current behavior.